sw/device/tests/dif/dif_padctrl_test.cc

```diff
@@ -0,0 +1,756 @@
+    TestEnableAttrFailureConflict(attr_field, kDifPadctrlPadDio);
+  }
+}
+
+TEST_F(DifPadctrlTest, WriteAttrFailureConflict) {
+  for (auto attr_field : mio_attr_fields_) {
+    TestWriteAttrFailureConflict(attr_field, kDifPadctrlPadMio);
+  }
+
+  for (auto attr_field : dio_attr_fields_) {
+    TestWriteAttrFailureConflict(attr_field, kDifPadctrlPadDio);
+  }
+}
+
+TEST_F(DifPadctrlTest, EnableAttrFailureInvalidPad) {
+  DeviceInit();
+
+  EXPECT_EQ(
+      dif_padctrl_enable_attrs(&dif_padctrl_, kDifPadctrlPadMio,
+                               PADCTRL_PARAM_NMIOPADS, kDifPadctrlAttrPullUp),
+      kDifPadctrlChangeBadArg);
+
+  EXPECT_EQ(
+      dif_padctrl_enable_attrs(&dif_padctrl_, kDifPadctrlPadDio,
+                               PADCTRL_PARAM_NDIOPADS, kDifPadctrlAttrPullUp),
+      kDifPadctrlChangeBadArg);
+
+  EXPECT_EQ(dif_padctrl_enable_attrs(&dif_padctrl_, kDifPadctrlPadMio, -1,
+                                     kDifPadctrlAttrPullUp),
+            kDifPadctrlChangeBadArg);
+
+  EXPECT_EQ(dif_padctrl_enable_attrs(&dif_padctrl_, kDifPadctrlPadDio, -1,
+                                     kDifPadctrlAttrPullUp),
+            kDifPadctrlChangeBadArg);
+}
+
+TEST_F(DifPadctrlTest, DisableAttrFailureInvalidPad) {
+  DeviceInit();
+
+  EXPECT_EQ(
+      dif_padctrl_disable_attrs(&dif_padctrl_, kDifPadctrlPadMio,
+                                PADCTRL_PARAM_NMIOPADS, kDifPadctrlAttrPullUp),
+      kDifPadctrlChangeBadArg);
+
+  EXPECT_EQ(
+      dif_padctrl_disable_attrs(&dif_padctrl_, kDifPadctrlPadDio,
+                                PADCTRL_PARAM_NDIOPADS, kDifPadctrlAttrPullUp),
+      kDifPadctrlChangeBadArg);
+
+  EXPECT_EQ(dif_padctrl_disable_attrs(&dif_padctrl_, kDifPadctrlPadMio, -1,
+                                      kDifPadctrlAttrPullUp),
+            kDifPadctrlChangeBadArg);
+
+  EXPECT_EQ(dif_padctrl_disable_attrs(&dif_padctrl_, kDifPadctrlPadDio, -1,
+                                      kDifPadctrlAttrPullUp),
+            kDifPadctrlChangeBadArg);
+}
+
+TEST_F(DifPadctrlTest, WriteAttrFailureInvalidPad) {
+  DeviceInit();
+
+  EXPECT_EQ(
+      dif_padctrl_write_attrs(&dif_padctrl_, kDifPadctrlPadMio,
+                              PADCTRL_PARAM_NMIOPADS, kDifPadctrlAttrPullUp),
+      kDifPadctrlChangeBadArg);
+
+  EXPECT_EQ(
+      dif_padctrl_write_attrs(&dif_padctrl_, kDifPadctrlPadDio,
+                              PADCTRL_PARAM_NDIOPADS, kDifPadctrlAttrPullUp),
+      kDifPadctrlChangeBadArg);
+
+  EXPECT_EQ(dif_padctrl_write_attrs(&dif_padctrl_, kDifPadctrlPadMio, -1,
+                                    kDifPadctrlAttrPullUp),
+            kDifPadctrlChangeBadArg);
+
+  EXPECT_EQ(dif_padctrl_write_attrs(&dif_padctrl_, kDifPadctrlPadDio, -1,
+                                    kDifPadctrlAttrPullUp),
+            kDifPadctrlChangeBadArg);
+}
+
+TEST_F(DifPadctrlTest, EnableAttrsFailureInvalidAttr) {
+  DeviceInit();
+
+  for (auto pad_kind : {kDifPadctrlPadMio, kDifPadctrlPadDio}) {
```

Comment:
I _feel_ like these want to be separate tests (for error messages' sake), but I don't feel too strongly here.

Reply:
I did wonder about this too, I was trying to balance very repetitive test code vs better test failure reports. As one test actually does quite a lot a single test failure error isn't hugely helpful and you need to dig around a bit anyway, so splitting these out into seperate MIO/DIO tests won't help you much.

We could split the tests out further so a single test failure more obviously points to the issue to solve but you'll end up with a huge number of tests and it's not clear there's a nice way to do that when you get down to per-pad tests anyway (do we really want a bunch of copy/pasted EnableAttrsFailureInvalidAttrMioPadX tests for example that require altering any time someone alters the number of pads? Maybe the testing framework has something for this kind of repetitive test generation I'm unaware of that makes this cleaner).